src/OpenTelemetry/SuppressInstrumentationScope.cs

```diff
@@ -15,13 +15,15 @@
 // </copyright>
 
 using System;
+using System.Diagnostics;
 using OpenTelemetry.Context;
 
 namespace OpenTelemetry
 {
     public sealed class SuppressInstrumentationScope : IDisposable
     {
         private static readonly RuntimeContextSlot<bool> Slot = RuntimeContext.RegisterSlot<bool>("otel.suppress_instrumentation");
+        private static readonly RuntimeContextSlot<ActivitySpanId?> ActivitySlot = RuntimeContext.RegisterSlot<ActivitySpanId?>("otel.suppress_instrumentation_activity");
```

Comment:
👀

Reply:
Even I wanted to avoid this allocation. Suppress is called in `OnStartActivity(Activity activity, object payload)` of `GrpcClientDiagnosticListener`, `OnStopActivity(Activity activity, object payload)` will not get called if we don't have this special case.

Reply:
Ugh... this is kinda a bugger, but I understand why you need it. Just thinking out loud, what if this were not a sealed class and instead we had a derived class for this use case of suppressing downstream instrumentation?

Reply:
@alanwest I feel like this is indicative of a bug in the design.

Am I understanding this flow correctly?

* Grpc activity starts. OnStart is called. Suppression is turned on.
* Http activity starts. OnStart skipped.
* Http activity stops. OnStop skipped.
* Grpc activity stops. OnStop is skipped?

We (somehow) need to make sure that last OnStop fires?

Reply:
@CodeBlanch , You're correct. This is the current behavior, Grpc activity Onstop is skipped.

Reply:
> @alanwest I feel like this is indicative of a bug in the design.

Either that or maybe this is not quite the right tool for the job. I think the SuppressInstrumentationScope works well from the standpoint of exporters, but from the standpoint of suppressing things within a specific trace I think this is where things get trickier.

I spoke with @reyang a little while back. I was questioning whether this was going to be the right tool for the job of suppressing child activities. I felt that the concept of a "leaf" span or as I believe you called it a "terminal" span may make more sense. Though, Reiley's concerns are valid in that the concept of a terminal span is specific to traces, and that in the future we'll want to be able to suppress downstream metrics and logs as well.

The fact that we're bringing in the notion of an Activity/SpanId into the SuppressInstrumentationScope brings me some pause for similar reasons, but all may not be lost...

Metrics and logs do not have a notion of start/stop like activities do, so I think the standard implementation of the `SuppressInstrumentationScope` would work for them. This is why I was pondering an derived version of the `SuppressInstrumentationScope` that was "Activity" aware.

Reply:
What if we make the `otel.suppress_instrumentation` slot a ref-count?
For example:
* The default value of the slot is `-1`, which means "do not suppress".
* Each `SuppressInstrumentationScope` will increment the slot value (no contention since it is async/thread local).
* For gRPC case, if slot value is `-1`, just suppress everything.
* For gPRC case, if the slot value is `0`, the `OnStart` adapter will increment the value if user opt-in to omit underlying operations; the `OnStop` adapter will decrement the value; adapter will use the value to decide if the processor should be called or not.

Reply:
I like where you're going @reyang, though I think there's a little more to this...

`OnStop` does not get invoked because of this check here
https://github.com/open-telemetry/opentelemetry-dotnet/blob/bced8cf08a4d458850a6c81a1936e21012589655/src/OpenTelemetry/Instrumentation/DiagnosticSourceListener.cs#L40-L45

This occurs above the level of the gRPC instrumentation and prevents all instrumentation operations from occurring `OnStart`, `OnStop`, `OnException`, `OnCustom`. Similarly, if this were a library that was instrumented with `ActivitySource` then the the check happens in the `ActivityListener`.

Reply:
See if this would work https://github.com/open-telemetry/opentelemetry-dotnet/pull/1202.
I would imagine the usage should be as simple as:

```csharp
public override void OnStartActivity(Activity activity, object payload)
{
    if (options.StopGoingAnyFurther)
    {
        SuppressInstrumentationScope.Enter();
    }
}
```